Document/0x05j-Testing-Resiliency-Against-Reverse-Engineering.md

```diff
@@ -14,60 +14,71 @@ In the following section, we list some common root detection methods you'll enco
+- MEETS_BASIC_INTEGRITY: Running on Device without Google Play Services: The app is running on a device that passed basic system integrity checks. This may occur because the device may be running an unrecognized version of Android, has an unlocked bootloader, or has not been certified by the manufacturer.
+- MEETS_STRONG_INTEGRITY: The app is running on an Android device equipped with Google Play Services technology, ensuring a strong system integrity such as a hardware-protected boot process.
+- MEETS_VIRTUAL_INTEGRITY: Running on Emulator with Google Play Services: In this scenario, the app is running in a simulated environment of an Android device that has Google Play Services. The emulator was checked and approved for system integrity and meets the necessary Android compatibility requirements.
 
-In broad terms, `basicIntegrity` gives you a signal about the general integrity of the device and its API. Many Rooted devices fail `basicIntegrity`, as do emulators, virtual devices, and devices with signs of tampering, such as API hooks.
+**API Errors:**
 
-On the other hand, `ctsProfileMatch` gives you a much stricter signal about the compatibility of the device. Only unmodified devices that have been certified by Google can pass `ctsProfileMatch`. Devices that will fail `ctsProfileMatch` include the following:
+The API can return local errors such as "APP_NOT_INSTALLED" and "APP_UID_MISMATCH", which can indicate a fraud attempt or attack. In addition, outdated Google Play Services or Play Store can also cause errors, and it is important to check these situations to ensure proper integrity verification functionality and to ensure the environment is not intentionally set up for an attack. You can find more details on the [official page](https://developer.android.com/google/play/integrity/error-codes).
 
-- Devices that fail `basicIntegrity`
-- Devices with an unlocked bootloader
-- Devices with a custom system image (custom ROM)
-- Devices for which the manufacturer didn't apply for, or pass, Google certification
-- Devices with a system image built directly from the Android Open Source Program source files
-- Devices with a system image distributed as part of a beta or developer preview program (including the Android Beta Program)
+**Best practices:**
 
-##### Recommendations when using `SafetyNetApi.attest`
+1. Although Play Integrity provides an additional layer of security for Android applications, it is not a single security solution. It is important to have complementary measures in place, such as input data validation, user authentication and anti-fraud protection.
+2. It is important to avoid requesting attests too frequently in order to minimize the impact on device resources. Furthermore, it is good practice to use the API only in situations where device integrity verification is truly necessary.

```

Comment:
```suggestion
2. It is important to avoid querying the Play Protect API too frequently in order to minimize the impact on device resources. Furthermore, it is good practice to use the API only in situations where device integrity verification is truly necessary.

```

Reply:
I fully agree.